<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Welcome to outrigger’s documentation! &#8212; Outrigger 1.0.0 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="top" title="Outrigger 1.0.0 documentation" href="#" />
    <link rel="next" title="Installation" href="installation.html" />
    <link rel="apple-touch-icon" sizes="57x57" href="/apple-icon-57x57.png">
    <link rel="apple-touch-icon" sizes="60x60" href="/apple-icon-60x60.png">
    <link rel="apple-touch-icon" sizes="72x72" href="/apple-icon-72x72.png">
    <link rel="apple-touch-icon" sizes="76x76" href="/apple-icon-76x76.png">
    <link rel="apple-touch-icon" sizes="114x114" href="/apple-icon-114x114.png">
    <link rel="apple-touch-icon" sizes="120x120" href="/apple-icon-120x120.png">
    <link rel="apple-touch-icon" sizes="144x144" href="/apple-icon-144x144.png">
    <link rel="apple-touch-icon" sizes="152x152" href="/apple-icon-152x152.png">
    <link rel="apple-touch-icon" sizes="180x180" href="/apple-icon-180x180.png">
    <link rel="icon" type="image/png" sizes="192x192"  href="/android-icon-192x192.png">
    <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="96x96" href="/favicon-96x96.png">
    <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
    <link rel="manifest" href="/manifest.json">
    <meta name="msapplication-TileColor" content="#ffffff">
    <meta name="msapplication-TileImage" content="/ms-icon-144x144.png">
    <meta name="theme-color" content="#ffffff">
    <!--[if IE]>
      <script type="text/javascript" src="_static/excanvas.js"></script>

    <![endif]-->

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="welcome-to-outrigger-s-documentation">
<h1>Welcome to outrigger&#8217;s documentation!<a class="headerlink" href="#welcome-to-outrigger-s-documentation" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="https://github.com/YeoLab/outrigger"><img alt="OutriggerLogo" src="http://yeolab.github.io/outrigger/_static/logo-400px.png" /></a></p>
<p><a class="reference external" href="https://travis-ci.org/YeoLab/outrigger"><img alt="BuildStatus" src="https://travis-ci.org/YeoLab/outrigger.svg?branch=master" /></a><a class="reference external" href="https://codecov.io/gh/YeoLab/outrigger"><img alt="codecov" src="https://codecov.io/gh/YeoLab/outrigger/branch/master/graph/badge.svg" /></a><a class="reference external" href="https://pypi.python.org/pypi/outrigger"><img alt="PyPIVersions" src="https://img.shields.io/pypi/v/outrigger.svg" /></a><a class="reference external" href="https://pypi.python.org/pypi/outrigger"><img alt="PythonVersionCompatibility" src="https://img.shields.io/pypi/pyversions/outrigger.svg" /></a></p>
</div>
<div class="section" id="fast-detection-and-accurate-calculation-of-alternative-splicing-with-outrigger">
<h1>Fast detection and accurate calculation of alternative splicing with <code class="docutils literal"><span class="pre">outrigger</span></code><a class="headerlink" href="#fast-detection-and-accurate-calculation-of-alternative-splicing-with-outrigger" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="https://github.com/YeoLab/outrigger">outrigger</a> is a program which uses junction reads from RNA seq data, and
a graph database to create a <em>de novo</em> alternative splicing annotation
with a graph database, and quantify percent spliced-in (Psi) of the
events.</p>
<ul class="simple">
<li>Free software: BSD license</li>
<li>Documentation is available here: <a class="reference external" href="http://yeolab.github.io/outrigger/">http://yeolab.github.io/outrigger/</a></li>
</ul>
<div class="section" id="features">
<h2>Features<a class="headerlink" href="#features" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Finds novel splicing events, including novel exons!
(<code class="docutils literal"><span class="pre">outrigger</span> <span class="pre">index</span></code>) from <code class="docutils literal"><span class="pre">.bam</span></code> files</li>
<li>(optional) Validates that exons have correct splice sites, e.g. GT/AG
and AT/AC for mammalian systems (<code class="docutils literal"><span class="pre">outrigger</span> <span class="pre">validate</span></code>)</li>
<li>Calculate &#8220;percent spliced-in&#8221; (Psi/Ψ) scores for all your samples
given the validated events (or the original events if you opted not
to validate) via <code class="docutils literal"><span class="pre">outrigger</span> <span class="pre">psi</span></code></li>
</ul>
<p><a class="reference external" href="http://yeolab.github.io/outrigger/_static/outrigger_overview-300ppi.png"><img alt="OutriggerOverview" src="http://yeolab.github.io/outrigger/_static/outrigger_overview-1x.png" /></a></p>
</div>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>To install <code class="docutils literal"><span class="pre">outrigger</span></code>, we recommend using the <a class="reference external" href="http://anaconda.org/">Anaconda Python
Distribution</a> and creating an environment.</p>
<p>You&#8217;ll want to add the <a class="reference external" href="https://bioconda.github.io/">bioconda</a> channel to make installing <a class="reference external" href="http://bedtools.readthedocs.io">bedtools</a> and its
Python wrapper, <a class="reference external" href="https://daler.github.io/pybedtools/">pybedtools</a> easy, as these programs are necessary for both
<code class="docutils literal"><span class="pre">outrigger</span> <span class="pre">index</span></code> and <code class="docutils literal"><span class="pre">outrigger</span> <span class="pre">validate</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">config</span> <span class="o">--</span><span class="n">add</span> <span class="n">channels</span> <span class="n">r</span>
<span class="n">conda</span> <span class="n">config</span> <span class="o">--</span><span class="n">add</span> <span class="n">channels</span> <span class="n">bioconda</span>
</pre></div>
</div>
<p>Create an environment called <code class="docutils literal"><span class="pre">outrigger-env</span></code>. Python 2.7, Python 3.4,
Python 3.5, and Python 3.6 are supported.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">create</span> <span class="o">--</span><span class="n">name</span> <span class="n">outrigger</span><span class="o">-</span><span class="n">env</span> <span class="n">outrigger</span>
</pre></div>
</div>
<p>Now activate that environment:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">source</span> <span class="n">activate</span> <span class="n">outrigger</span><span class="o">-</span><span class="n">env</span>
</pre></div>
</div>
<p>To check that it installed properly, try the command with the help
option (<code class="docutils literal"><span class="pre">-h</span></code>), <code class="docutils literal"><span class="pre">outrigger</span> <span class="pre">-h</span></code>. The output should look like this:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ outrigger -h
usage: outrigger [-h] [--version] {index,validate,psi} ...

outrigger (1.0.0dev). Calculate &quot;percent-spliced in&quot; (Psi) scores of
alternative splicing on a *de novo*, custom-built splicing index -- just for
you!

positional arguments:
  {index,validate,psi}  Sub-commands
    index               Build an index of splicing events using a graph
                        database on your junction reads and an annotation
    validate            Ensure that the splicing events found all have the
                        correct splice sites
    psi                 Calculate &quot;percent spliced-in&quot; (Psi) values using the
                        splicing event index built with &quot;outrigger index&quot;

optional arguments:
  -h, --help            show this help message and exit
  --version             show program&#39;s version number and exit
</pre></div>
</div>
<div class="section" id="bleeding-edge-code-from-github-here">
<h3>Bleeding edge code from Github (here)<a class="headerlink" href="#bleeding-edge-code-from-github-here" title="Permalink to this headline">¶</a></h3>
<p>For advanced users, if you have <a class="reference external" href="https://git-scm.com">git</a> and
<a class="reference external" href="https://www.continuum.io/downloads">Anaconda Python</a> installed, you
can:</p>
<ol class="arabic simple">
<li>Clone this repository</li>
<li>Change into that directory</li>
<li>Create an environment named <code class="docutils literal"><span class="pre">outrigger-env</span></code> with the necessary packages
from Anaconda and the Python Package Index (PyPI).</li>
<li>Activate the environment</li>
</ol>
<p>These steps are shown in code below.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">git</span> <span class="n">clone</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">YeoLab</span><span class="o">/</span><span class="n">outrigger</span><span class="o">.</span><span class="n">git</span>
<span class="n">cd</span> <span class="n">outrigger</span>
<span class="n">conda</span> <span class="n">env</span> <span class="n">create</span> <span class="o">--</span><span class="n">file</span> <span class="n">environment</span><span class="o">.</span><span class="n">yml</span>
<span class="n">source</span> <span class="n">activate</span> <span class="n">outrigger</span><span class="o">-</span><span class="n">env</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="quick-start">
<h2>Quick start<a class="headerlink" href="#quick-start" title="Permalink to this headline">¶</a></h2>
<p>If you just want to know how to run this on your data with the default
parameters, start here. Let&#8217;s say you performed your alignment in the
folder called <code class="docutils literal"><span class="pre">~/projects/tasic2016/analysis/tasic2016_v1</span></code>, and that&#8217;s
where your <code class="docutils literal"><span class="pre">SJ.out.tab</span></code> files from the STAR aligner are (they&#8217;re
output into the same folder as the <code class="docutils literal"><span class="pre">.bam</span></code> files). First you&#8217;ll need to
change directories to that folder with <code class="docutils literal"><span class="pre">cd</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="o">~/</span><span class="n">projects</span><span class="o">/</span><span class="n">tasic2016</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">tasic2016_v1</span>
</pre></div>
</div>
<p>Then you need find all alternative splicing events, which you do by
running <code class="docutils literal"><span class="pre">outrigger</span> <span class="pre">index</span></code> on the splice junction files and the gtf.
Here is an example command:</p>
<div class="section" id="input-sj-out-tab-files">
<h3>Input: <code class="docutils literal"><span class="pre">.SJ.out.tab</span></code> files<a class="headerlink" href="#input-sj-out-tab-files" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">outrigger</span> <span class="n">index</span> <span class="o">--</span><span class="n">sj</span><span class="o">-</span><span class="n">out</span><span class="o">-</span><span class="n">tab</span> <span class="o">*</span><span class="n">SJ</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">tab</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">ps</span><span class="o">-</span><span class="n">yeolab</span><span class="o">/</span><span class="n">genomes</span><span class="o">/</span><span class="n">mm10</span><span class="o">/</span><span class="n">gencode</span><span class="o">/</span><span class="n">m10</span><span class="o">/</span><span class="n">gencode</span><span class="o">.</span><span class="n">vM10</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">gtf</span>
</pre></div>
</div>
</div>
<div class="section" id="input-bam-files">
<h3>Input: <code class="docutils literal"><span class="pre">.bam</span></code> files<a class="headerlink" href="#input-bam-files" title="Permalink to this headline">¶</a></h3>
<p>If you&#8217;re using <code class="docutils literal"><span class="pre">.bam</span></code> files instead of <code class="docutils literal"><span class="pre">SJ.out.tab</span></code> files, never despair!
Below is an example command. Keep in mind that for this program to work, the
events must be sorted and indexed.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">outrigger</span> <span class="n">index</span> <span class="o">--</span><span class="n">bam</span> <span class="o">*</span><span class="nb">sorted</span><span class="o">.</span><span class="n">bam</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">ps</span><span class="o">-</span><span class="n">yeolab</span><span class="o">/</span><span class="n">genomes</span><span class="o">/</span><span class="n">mm10</span><span class="o">/</span><span class="n">gencode</span><span class="o">/</span><span class="n">m10</span><span class="o">/</span><span class="n">gencode</span><span class="o">.</span><span class="n">vM10</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">gtf</span>
</pre></div>
</div>
<p>Next, you&#8217;ll want to validate that the splicing events you found follow
biological rules, such as being containing GT/AG (mammalian major
spliceosome) or AT/AC (mammalian minor splicesome) sequences. To do
that, you&#8217;ll need to provide the genome name (e.g. <code class="docutils literal"><span class="pre">mm10</span></code>) and the
genome sequences. An example command is below:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">outrigger</span> <span class="n">validate</span> <span class="o">--</span><span class="n">genome</span> <span class="n">mm10</span> \
    <span class="o">--</span><span class="n">fasta</span> <span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">ps</span><span class="o">-</span><span class="n">yeolab</span><span class="o">/</span><span class="n">genomes</span><span class="o">/</span><span class="n">mm10</span><span class="o">/</span><span class="n">GRCm38</span><span class="o">.</span><span class="n">primary_assembly</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">fa</span>
</pre></div>
</div>
<p>Finally, you can calculate percent spliced in (Psi) of your splicing
events! Thankfully this is very easy:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">outrigger</span> <span class="n">psi</span>
</pre></div>
</div>
<p>It should be noted that ALL of these commands should be performed in the
same directory, so no moving.</p>
</div>
<div class="section" id="quick-start-summary">
<h3>Quick start summary<a class="headerlink" href="#quick-start-summary" title="Permalink to this headline">¶</a></h3>
<p>Here is a summary the commands in the order you would use them for
outrigger!</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="o">~/</span><span class="n">projects</span><span class="o">/</span><span class="n">tasic2016</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">tasic2016_v1</span>
<span class="n">outrigger</span> <span class="n">index</span> <span class="o">--</span><span class="n">sj</span><span class="o">-</span><span class="n">out</span><span class="o">-</span><span class="n">tab</span> <span class="o">*</span><span class="n">SJ</span><span class="o">.</span><span class="n">out</span><span class="o">.</span><span class="n">tab</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">ps</span><span class="o">-</span><span class="n">yeolab</span><span class="o">/</span><span class="n">genomes</span><span class="o">/</span><span class="n">mm10</span><span class="o">/</span><span class="n">gencode</span><span class="o">/</span><span class="n">m10</span><span class="o">/</span><span class="n">gencode</span><span class="o">.</span><span class="n">vM10</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">gtf</span>
<span class="n">outrigger</span> <span class="n">validate</span> <span class="o">--</span><span class="n">genome</span> <span class="n">mm10</span> \
    <span class="o">--</span><span class="n">fasta</span> <span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">ps</span><span class="o">-</span><span class="n">yeolab</span><span class="o">/</span><span class="n">genomes</span><span class="o">/</span><span class="n">mm10</span><span class="o">/</span><span class="n">GRCm38</span><span class="o">.</span><span class="n">primary_assembly</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">fa</span>
<span class="n">outrigger</span> <span class="n">psi</span>
</pre></div>
</div>
<p>This will create a folder called <code class="docutils literal"><span class="pre">outrigger_output</span></code>, which at the end
should look like the one below. Each file and folder is annotated with which command
produced it.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ tree outrigger_output
outrigger_output..........................................................index
├── index.................................................................index
│   ├── gtf...............................................................index
│   │   ├── gencode.vM10.annotation.gtf...................................index
│   │   ├── gencode.vM10.annotation.gtf.db................................index
│   │   └── novel_exons.gtf...............................................index
│   ├── exon_direction_junction_triples.csv...............................index
│   ├── mxe...............................................................index
│   │   ├── event.bed.....................................................index
│   │   ├── events.csv....................................................index
│   │   ├── exon1.bed.....................................................index
│   │   ├── exon2.bed.....................................................index
│   │   ├── exon3.bed.....................................................index
│   │   ├── exon4.bed.....................................................index
│   │   ├── intron.bed....................................................index
│   │   ├── splice_sites.csv...........................................validate
│   │   └── validated..................................................validate
│   │       └── events.csv.............................................validate
│   └── se................................................................index
│       ├── event.bed.....................................................index
│       ├── events.csv....................................................index
│       ├── exon1.bed.....................................................index
│       ├── exon2.bed.....................................................index
│       ├── exon3.bed.....................................................index
│       ├── intron.bed....................................................index
│       ├── splice_sites.csv...........................................validate
│       └── validated..................................................validate
│           └── events.csv.............................................validate
├── junctions.............................................................index
│   ├── metadata.csv......................................................index
│   └── reads.csv.........................................................index
└── psi.....................................................................psi
    ├── mxe.................................................................psi
    |   ├── psi.csv.........................................................psi
    │   └── summary.csv.....................................................psi
    ├── outrigger_psi.csv...................................................psi
    └── se..................................................................psi
        ├── psi.csv.........................................................psi
        └── summary.csv.....................................................psi

10 directories, 26 files
</pre></div>
</div>
<p>Contents:</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="subcommands/outrigger_index.html"><code class="docutils literal"><span class="pre">index</span></code>: Build a <em>de novo</em> splicing annotation index of events custom to <strong>your</strong> data</a></li>
<li class="toctree-l2"><a class="reference internal" href="subcommands/outrigger_validate.html"><code class="docutils literal"><span class="pre">validate</span></code>: Check that the found exons are real</a></li>
<li class="toctree-l2"><a class="reference internal" href="subcommands/outrigger_psi.html"><code class="docutils literal"><span class="pre">psi</span></code>: Calculate percent spliced-in (Psi/Ψ) scores for your data from the splicing events you created</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#types-of-contributions">Types of Contributions</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#tips-for-writing-documentation">Tips for writing documentation</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#get-started">Get Started!</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#pull-request-guidelines">Pull Request Guidelines</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#tips">Tips</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a><ul>
<li class="toctree-l2"><a class="reference internal" href="authors.html#development-lead">Development Lead</a></li>
<li class="toctree-l2"><a class="reference internal" href="authors.html#contributors">Contributors</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a><ul>
<li class="toctree-l2"><a class="reference internal" href="history.html#v1-0-0-april-3rd-2017">v1.0.0 (April 3rd, 2017)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#v0-2-9-november-11th-2016">v0.2.9 (November 11th, 2016)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#v0-2-8-october-23rd-2016">v0.2.8 (October 23rd, 2016)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#v0-2-7-october-23rd-2016">v0.2.7 (October 23rd, 2016)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#v0-2-6-september-15th-2016">v0.2.6 (September 15th, 2016)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#v0-2-5-september-14th-2016">v0.2.5 (September 14th, 2016)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#v0-2-4-september-14th-2016">v0.2.4 (September 14th, 2016)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#v0-2-3-september-13th-2016">v0.2.3 (September 13th, 2016)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#v0-2-2-september-12th-2016">v0.2.2 (September 12th, 2016)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#v0-2-1-september-12th-2016">v0.2.1 (September 12th, 2016)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#v0-2-0-september-9th-2016">v0.2.0 (September 9th, 2016)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#v0-1-0-may-25-2016">v0.1.0 (May 25, 2016)</a></li>
</ul>
</li>
</ul>
</div>
</div>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="#">
              <img class="logo" src="_static/logo-150px.png" alt="Logo"/>
            </a></p><h3>About outrigger</h3>
<p>
  outrigger is an RNA-seq analysis tool that helps
    you be confident in your alternative exon inclusion calculations.
</p><h3><a href="#">Table Of Contents</a></h3>

<ul>
  <li><a href="#">Home</a></li>
  <li><a href="contents.html">Contents</a></li>
  <li><a href="installation.html">Install</a></li>
  <li><a href="Usage.html">Usage</a></li>
  <li><a href="subcommands/outrigger_index.html"><code>index</code>: Detect exons</a></li>
  <li><a href="subcommands/outrigger_validate.html"><code>validate</code>: Remove non-canonical splice sites</a></li>
  <li><a href="subcommands/outrigger_psi.html"><code>psi</code>: Quantify exon inclusion</a></li>
  <li><a href="history.html">Changelog</a></li>
  <li><a href="license.html">License</a></li>
</ul>
  <hr>
  <ul>
<li><a class="reference internal" href="#">Welcome to outrigger&#8217;s documentation!</a></li>
<li><a class="reference internal" href="#fast-detection-and-accurate-calculation-of-alternative-splicing-with-outrigger">Fast detection and accurate calculation of alternative splicing with <code class="docutils literal"><span class="pre">outrigger</span></code></a><ul>
<li><a class="reference internal" href="#features">Features</a></li>
<li><a class="reference internal" href="#installation">Installation</a><ul>
<li><a class="reference internal" href="#bleeding-edge-code-from-github-here">Bleeding edge code from Github (here)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#quick-start">Quick start</a><ul>
<li><a class="reference internal" href="#input-sj-out-tab-files">Input: <code class="docutils literal"><span class="pre">.SJ.out.tab</span></code> files</a></li>
<li><a class="reference internal" href="#input-bam-files">Input: <code class="docutils literal"><span class="pre">.bam</span></code> files</a></li>
<li><a class="reference internal" href="#quick-start-summary">Quick start summary</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>

<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015-2017, Olga Botvinnik.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.4.8</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.9</a>
      
      |
      <a href="_sources/index.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>