<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Outrigger &mdash; BED annotator 0.2.9 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.2.9',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="BED annotator 0.2.9 documentation" href="index.html" />
    <link rel="next" title="Installation" href="installation.html" />
    <link rel="prev" title="Welcome to outrigger‘s documentation!" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to outrigger‘s documentation!"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">BED annotator 0.2.9 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="outrigger">
<h1>Outrigger<a class="headerlink" href="#outrigger" title="Permalink to this headline">¶</a></h1>
<p><img alt="OutriggerLogo" src="https://raw.githubusercontent.com/YeoLab/outrigger/master/logo/logo_v1.png" /></p>
<p><a class="reference external" href="https://travis-ci.org/YeoLab/outrigger"><img alt="BuildStatus" src="https://travis-ci.org/YeoLab/outrigger.svg?branch=master" /></a><a class="reference external" href="https://codecov.io/gh/YeoLab/outrigger"><img alt="codecov" src="https://codecov.io/gh/YeoLab/outrigger/branch/master/graph/badge.svg" /></a><a class="reference external" href="https://pypi.python.org/pypi/outrigger"><img alt="PyPI versions" src="https://img.shields.io/pypi/v/outrigger.svg" /></a><a class="reference external" href="https://pypi.python.org/pypi/outrigger"><img alt="PythonVersionCompatibility" src="https://img.shields.io/pypi/pyversions/outrigger.svg" /></a></p>
<p>Outrigger is a program which uses junction reads from RNA seq data, and
a graph database to create a <em>de novo</em> alternative splicing annotation
with a graph database, and quantify percent spliced-in (Psi) of the
events.</p>
<ul class="simple">
<li>Free software: BSD license</li>
</ul>
<div class="section" id="features">
<h2>Features<a class="headerlink" href="#features" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Finds novel splicing events, including novel exons!
(<code class="docutils literal"><span class="pre">outrigger</span> <span class="pre">index</span></code>) from <code class="docutils literal"><span class="pre">.bam</span></code> files</li>
<li>(optional) Validates that exons have correct splice sites, e.g. GT/AG
and AT/AC for mammalian systems (<code class="docutils literal"><span class="pre">outrigger</span> <span class="pre">validate</span></code>)</li>
<li>Calculate &#8220;percent spliced-in&#8221; (Psi/Ψ) scores for all your samples
given the validated events (or the original events if you opted not
to validate)</li>
</ul>
</div>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>To install <code class="docutils literal"><span class="pre">outrigger</span></code>, we recommend using the <a class="reference external" href="http://anaconda.org/">Anaconda Python
Distribution</a> and creating an environment.</p>
<p>You&#8217;ll want to add the <code class="docutils literal"><span class="pre">`bioconda</span></code> &lt;<a class="reference external" href="https://bioconda.github.io/">https://bioconda.github.io/</a>&gt;`__
channel to make installing <code class="docutils literal"><span class="pre">`bedtools</span></code> &lt;bedtools.readthedocs.io&gt;`__
and its Python wrapper,
<code class="docutils literal"><span class="pre">`pybedtools</span></code> &lt;<a class="reference external" href="https://daler.github.io/pybedtools/">https://daler.github.io/pybedtools/</a>&gt;`__ easy.</p>
<div class="highlight-python"><div class="highlight"><pre>conda config --add channels r
conda config --add channels bioconda
</pre></div>
</div>
<p>Create an environment called <code class="docutils literal"><span class="pre">outrigger-env</span></code>. Python 2.7, Python 3.4,
and Python 3.5 are supported.</p>
<div class="highlight-python"><div class="highlight"><pre>conda create -n outrigger-env pandas pybedtools gffutils biopython bedtools joblib
</pre></div>
</div>
<p>Now activate that environment using <code class="docutils literal"><span class="pre">source</span> <span class="pre">activate</span> <span class="pre">outrigger-env</span></code>
and install <code class="docutils literal"><span class="pre">outrigger</span></code> from PyPI, using <code class="docutils literal"><span class="pre">pip</span></code>:</p>
<div class="highlight-python"><div class="highlight"><pre>source activate outrigger-env
pip install outrigger
</pre></div>
</div>
<p>To check that it installed properly, try the command with the help
option (<code class="docutils literal"><span class="pre">-h</span></code>), <code class="docutils literal"><span class="pre">outrigger</span> <span class="pre">-h</span></code>. The output should look like this:</p>
<div class="highlight-python"><div class="highlight"><pre>$ outrigger -h
usage: outrigger [-h] {index,validate,psi} ...

Calculate &quot;percent-spliced in&quot; (Psi) scores of alternative splicing on a *de
novo*, custom-built splicing index

positional arguments:
  {index,validate,psi}  Sub-commands
    index               Build an index of splicing events using a graph
                        database on your junction reads and an annotation
    validate            Ensure that the splicing events found all have the
                        correct splice sites
    psi                 Calculate &quot;percent spliced-in&quot; (Psi) values using the
                        splicing event index built with &quot;outrigger index&quot;

optional arguments:
  -h, --help            show this help message and exit
</pre></div>
</div>
<div class="section" id="bleeding-edge-code-from-github-here">
<h3>Bleeding edge code from Github (here)<a class="headerlink" href="#bleeding-edge-code-from-github-here" title="Permalink to this headline">¶</a></h3>
<p>For advanced users, if you have <a class="reference external" href="https://git-scm.com/">git</a> and
<a class="reference external" href="https://www.continuum.io/downloads">Anaconda Python</a> installed, you
can:</p>
<ol class="arabic simple">
<li>Clone this repository</li>
<li>Change into that directory</li>
<li>Create an environment with the necessary packages from Anaconda</li>
<li>Activate the environment</li>
<li>Install remaining packages from PyPI
(<code class="docutils literal"><span class="pre">`graphlite</span></code> &lt;<a class="reference external" href="https://github.com/eugene-eeo/graphlite">https://github.com/eugene-eeo/graphlite</a>&gt;`__ is only
available on PyPI, not as a <code class="docutils literal"><span class="pre">conda</span></code> package)</li>
<li>Install this package</li>
</ol>
<p>These steps are shown in code below.</p>
<div class="highlight-python"><div class="highlight"><pre>git clone git@github.com:YeoLab/outrigger
cd outrigger
conda create --name outrigger --yes --file conda_requirements.txt --channel bioconda
source activate outrigger
pip install -r requirements.txt
pip install .
</pre></div>
</div>
</div>
</div>
<div class="section" id="quick-start">
<h2>Quick start<a class="headerlink" href="#quick-start" title="Permalink to this headline">¶</a></h2>
<p>If you just want to know how to run this on your data with the default
parameters, start here. Let&#8217;s say you performed your alignment in the
folder called <code class="docutils literal"><span class="pre">~/projects/tasic2016/analysis/tasic2016_v1</span></code>, and that&#8217;s
where your <code class="docutils literal"><span class="pre">SJ.out.tab</span></code> files from the STAR aligner are (they&#8217;re
output into the same folder as the <code class="docutils literal"><span class="pre">.bam</span></code> files). First you&#8217;ll need to
change directories to that folder with <code class="docutils literal"><span class="pre">cd</span></code>.</p>
<div class="highlight-python"><div class="highlight"><pre>cd ~/projects/tasic2016/analysis/tasic2016_v1
</pre></div>
</div>
<p>Then you need find all alternative splicing events, which you do by
running <code class="docutils literal"><span class="pre">outrigger</span> <span class="pre">index</span></code> on the splice junction files and the gtf.
Here is an example command:</p>
<div class="highlight-python"><div class="highlight"><pre>outrigger index --sj-out-tab *SJ.out.tab \
    --gtf /projects/ps-yeolab/genomes/mm10/gencode/m10/gencode.vM10.annotation.gtf
</pre></div>
</div>
<p>Next, you&#8217;ll want to validate that the splicing events you found follow
biological rules, such as being containing GT/AG (mammalian major
spliceosome) or AT/AC (mammalian minor splicesome) sequences. To do
that, you&#8217;ll need to provide the genome name (e.g. <code class="docutils literal"><span class="pre">mm10</span></code>) and the
genome sequences. An example command is below:</p>
<div class="highlight-python"><div class="highlight"><pre>outrigger validate --genome mm10 \
    --fasta /projects/ps-yeolab/genomes/mm10/GRCm38.primary_assembly.genome.fa
</pre></div>
</div>
<p>Finally, you can calculate percent spliced in (Psi) of your splicing
events! Thankfully this is very easy:</p>
<div class="highlight-python"><div class="highlight"><pre>outrigger psi
</pre></div>
</div>
<p>It should be noted that ALL of these commands should be performed in the
same directory, so no moving.</p>
<div class="section" id="quick-start-summary">
<h3>Quick start summary<a class="headerlink" href="#quick-start-summary" title="Permalink to this headline">¶</a></h3>
<p>Here is a summary the commands in the order you would use them for
outrigger!</p>
<div class="highlight-python"><div class="highlight"><pre>cd ~/projects/tasic2016/analysis/tasic2016_v1
outrigger index --sj-out-tab *SJ.out.tab \
    --gtf /projects/ps-yeolab/genomes/mm10/gencode/m10/gencode.vM10.annotation.gtf
outrigger validate --genome mm10 \
    --fasta /projects/ps-yeolab/genomes/mm10/GRCm38.primary_assembly.genome.fa
outrigger psi
</pre></div>
</div>
<p>This will create a folder called <code class="docutils literal"><span class="pre">outrigger_output</span></code>, which at the end
should look like this:</p>
<div class="highlight-python"><div class="highlight"><pre>$ tree outrigger_output
outrigger_output
├── index
│   ├── gtf
│   │   ├── gencode.vM10.annotation.gtf
│   │   ├── gencode.vM10.annotation.gtf.db
│   │   └── novel_exons.gtf
│   ├── junction_exon_direction_triples.csv
│   ├── mxe
│   │   ├── events.csv
│   │   ├── exon1.bed
│   │   ├── exon2.bed
│   │   ├── exon3.bed
│   │   ├── exon4.bed
│   │   ├── splice_sites.csv
│   │   └── validated
│   │       └── events.csv
│   └── se
│       ├── events.csv
│       ├── exon1.bed
│       ├── exon2.bed
│       ├── exon3.bed
│       ├── splice_sites.csv
│       └── validated
│           └── events.csv
├── junctions
│   ├── metadata.csv
│   └── reads.csv
└── psi
    ├── mxe
    │   └── psi.csv
    ├── outrigger_psi.csv
    └── se
        └── psi.csv

10 directories, 22 files
</pre></div>
</div>
</div>
</div>
<div class="section" id="for-developers">
<h2>For Developers<a class="headerlink" href="#for-developers" title="Permalink to this headline">¶</a></h2>
<div class="section" id="how-to-run-with-the-python-debugger">
<h3>How to run with the Python debugger<a class="headerlink" href="#how-to-run-with-the-python-debugger" title="Permalink to this headline">¶</a></h3>
<p>How to run the code with the Python debugger. To run the command line
functions such that when they break, you jump into the <code class="docutils literal"><span class="pre">pdb</span></code> (Python
debugger), here is the code:</p>
<div class="highlight-python"><div class="highlight"><pre>python -m pdb outrigger/commandline.py index \
--sj-out-tab outrigger/test_data/tasic2016/unprocessed/sj_out_tab/* \
    --gtf outrigger/test_data/tasic2016/unprocessed/gtf/gencode.vM10.annotation.snap25.myl6.gtf
</pre></div>
</div>
<p>Notice that you replace <code class="docutils literal"><span class="pre">outrigger</span></code> with
<code class="docutils literal"><span class="pre">python</span> <span class="pre">-m</span> <span class="pre">pdb</span> <span class="pre">outrigger/commandline.py</span></code>, which is relative to this
github directory.</p>
</div>
<div class="section" id="how-to-run-the-tests">
<h3>How to run the tests<a class="headerlink" href="#how-to-run-the-tests" title="Permalink to this headline">¶</a></h3>
<p>If you want to run the tests without calculating what percentage of
lines are covered in the test suite, run</p>
<div class="highlight-python"><div class="highlight"><pre>make test
</pre></div>
</div>
<p>If you want to run the tests and see which lines are covered by tests
and get an overall percentage of test coverage, run</p>
<div class="highlight-python"><div class="highlight"><pre>make coverage
</pre></div>
</div>
<p>If you want to run an example with ENSEMBL GTF files, do:</p>
<div class="highlight-python"><div class="highlight"><pre>make arabdopsis
</pre></div>
</div>
<p>By default, Travis-CI does all three:</p>
<div class="highlight-python"><div class="highlight"><pre>script:
- make coverage
- make lint
- make arabdopsis
</pre></div>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Outrigger</a><ul>
<li><a class="reference internal" href="#features">Features</a></li>
<li><a class="reference internal" href="#installation">Installation</a><ul>
<li><a class="reference internal" href="#bleeding-edge-code-from-github-here">Bleeding edge code from Github (here)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#quick-start">Quick start</a><ul>
<li><a class="reference internal" href="#quick-start-summary">Quick start summary</a></li>
</ul>
</li>
<li><a class="reference internal" href="#for-developers">For Developers</a><ul>
<li><a class="reference internal" href="#how-to-run-with-the-python-debugger">How to run with the Python debugger</a></li>
<li><a class="reference internal" href="#how-to-run-the-tests">How to run the tests</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to <code class="docutils literal"><span class="pre">outrigger</span></code>&#8216;s documentation!</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="installation.html"
                        title="next chapter">Installation</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/readme.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to outrigger‘s documentation!"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">BED annotator 0.2.9 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Olga Botvinnik.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>