<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>outrigger.index.adjacencies &mdash; BED annotator 0.2.9 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '0.2.9',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <link rel="top" title="BED annotator 0.2.9 documentation" href="../../../index.html" />
    <link rel="up" title="Module code" href="../../index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">BED annotator 0.2.9 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for outrigger.index.adjacencies</h1><div class="highlight"><pre>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Find exons adjacent to junctions</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">sqlite3</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">gffutils</span>
<span class="kn">from</span> <span class="nn">gffutils.helpers</span> <span class="kn">import</span> <span class="n">merge_attributes</span>
<span class="kn">import</span> <span class="nn">joblib</span>

<span class="kn">from</span> <span class="nn">..common</span> <span class="kn">import</span> <span class="n">JUNCTION_ID</span><span class="p">,</span> <span class="n">EXON_START</span><span class="p">,</span> <span class="n">EXON_STOP</span><span class="p">,</span> <span class="n">CHROM</span><span class="p">,</span> <span class="n">STRAND</span><span class="p">,</span> \
    <span class="n">ORDER_BY</span>
<span class="kn">from</span> <span class="nn">..io.gtf</span> <span class="kn">import</span> <span class="n">transform</span><span class="p">,</span> <span class="n">maybe_analyze</span>
<span class="kn">from</span> <span class="nn">..region</span> <span class="kn">import</span> <span class="n">Region</span><span class="p">,</span> <span class="n">STRANDS</span>
<span class="kn">from</span> <span class="nn">..util</span> <span class="kn">import</span> <span class="n">done</span><span class="p">,</span> <span class="n">progress</span>


<span class="k">with</span> <span class="n">warnings</span><span class="o">.</span><span class="n">catch_warnings</span><span class="p">():</span>
    <span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s">&quot;ignore&quot;</span><span class="p">)</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>


<span class="n">UPSTREAM</span> <span class="o">=</span> <span class="s">&#39;upstream&#39;</span>
<span class="n">DOWNSTREAM</span> <span class="o">=</span> <span class="s">&#39;downstream&#39;</span>
<span class="n">DIRECTIONS</span> <span class="o">=</span> <span class="n">UPSTREAM</span><span class="p">,</span> <span class="n">DOWNSTREAM</span>
<span class="n">NOVEL_EXON</span> <span class="o">=</span> <span class="s">&#39;novel_exon&#39;</span>
<span class="n">OUTRIGGER_DE_NOVO</span> <span class="o">=</span> <span class="s">&#39;outrigger_de_novo&#39;</span>
<span class="n">MAX_DE_NOVO_EXON_LENGTH</span> <span class="o">=</span> <span class="mi">100</span>


<span class="k">def</span> <span class="nf">_unify_strand</span><span class="p">(</span><span class="n">strand1</span><span class="p">,</span> <span class="n">strand2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;If strands are equal, return the strand, otherwise return None&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">strand1</span> <span class="o">!=</span> <span class="n">strand2</span><span class="p">:</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="s">&#39;.&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="n">strand1</span>
    <span class="k">return</span> <span class="n">strand</span>


<span class="k">def</span> <span class="nf">_exons_from_neighboring_junctions</span><span class="p">(</span><span class="n">junction</span><span class="p">,</span> <span class="n">neighbors</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s">&#39;left&#39;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns (chrom, start, stop, strand) of neighboring exons from junctions</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    junction : outrigger.Region</span>
<span class="sd">        A single junction whose neighboring exons you want</span>
<span class="sd">    neighbors : pandas.DataFrame</span>
<span class="sd">        All junctions which are within</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    exons : pandas.Series</span>
<span class="sd">        A column containing (chrom, start, stop, strand) for each detected exon</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">neighbors</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">side</span> <span class="o">==</span> <span class="s">&#39;left&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">neighbors</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span>
            <span class="n">x</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">stop</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">junction</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>
            <span class="n">_unify_strand</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span> <span class="n">junction</span><span class="o">.</span><span class="n">strand</span><span class="p">)),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">side</span> <span class="o">==</span> <span class="s">&#39;right&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">neighbors</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span>
            <span class="n">x</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">junction</span><span class="o">.</span><span class="n">stop</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>
            <span class="n">_unify_strand</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span> <span class="n">junction</span><span class="o">.</span><span class="n">strand</span><span class="p">)),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">_neighboring_exons</span><span class="p">(</span><span class="n">junction</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s">&#39;left&#39;</span><span class="p">,</span>
                       <span class="n">max_de_novo_exon_length</span><span class="o">=</span><span class="n">MAX_DE_NOVO_EXON_LENGTH</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Get either the left or right neighbors of a particular junction</span>

<span class="sd">    Used to find novel exons between junctions. Not part of the</span>
<span class="sd">    ExonJunctionAdjacencies object so it can be parallelized with joblib.</span>
<span class="sd">    Internal function</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    junction : outrigger.Region</span>
<span class="sd">        A Region object with .start and .stop</span>
<span class="sd">    df : pandas.DataFrame</span>
<span class="sd">        A data table with &quot;start&quot; and &quot;stop&quot; columns of all junctions</span>
<span class="sd">    side : &#39;left&#39; | &#39;right&#39;</span>
<span class="sd">        Specifies whether you want the left or right side neighbors of a</span>
<span class="sd">        junction (not strand-specific)</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    neighboring_exon_locations : pandas.Series</span>
<span class="sd">        A column containing (chrom, start, stop, strand) for each detected exon</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">side</span> <span class="o">==</span> <span class="s">&#39;left&#39;</span><span class="p">:</span>
        <span class="n">rows</span> <span class="o">=</span> <span class="n">junction</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">df</span><span class="o">.</span><span class="n">stop</span>
    <span class="k">elif</span> <span class="n">side</span> <span class="o">==</span> <span class="s">&#39;right&#39;</span><span class="p">:</span>
        <span class="n">rows</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">junction</span><span class="o">.</span><span class="n">stop</span>
    <span class="n">rows</span> <span class="o">=</span> <span class="n">rows</span><span class="p">[</span><span class="n">rows</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>
    <span class="n">rows</span> <span class="o">=</span> <span class="n">rows</span><span class="p">[</span><span class="n">rows</span> <span class="o">&lt;=</span> <span class="n">max_de_novo_exon_length</span><span class="p">]</span>
    <span class="n">neighboring_junctions</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">rows</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>

    <span class="n">neighboring_exon_locations</span> <span class="o">=</span> <span class="n">_exons_from_neighboring_junctions</span><span class="p">(</span>
        <span class="n">junction</span><span class="p">,</span> <span class="n">neighboring_junctions</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="n">side</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">neighboring_exon_locations</span>


<div class="viewcode-block" id="is_there_an_exon_here"><a class="viewcode-back" href="../../../outrigger.index.adjacencies.html#outrigger.index.adjacencies.is_there_an_exon_here">[docs]</a><span class="k">def</span> <span class="nf">is_there_an_exon_here</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">junction1</span><span class="p">,</span> <span class="n">junction2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Check if there could be an exon between these two junctions</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    junction{1,2} : outrigger.Region</span>
<span class="sd">        Outrigger.Region objects</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    start, stop : (int, int) or (False, False)</span>
<span class="sd">        Start and stop of the new exon if it exists, else False, False</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">junction1</span><span class="o">.</span><span class="n">overlaps</span><span class="p">(</span><span class="n">junction2</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">False</span><span class="p">,</span> <span class="bp">False</span>

    <span class="c"># These are junction start/stops, not exon start/stops</span>
    <span class="c"># Move one nt upstream of starts for exon stops,</span>
    <span class="c"># and one nt downstream of stops for exon starts.</span>
    <span class="n">option1</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">junction1</span><span class="o">.</span><span class="n">stop</span> <span class="o">-</span> <span class="n">junction2</span><span class="o">.</span><span class="n">start</span><span class="p">)</span> \
        <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_de_novo_exon_length</span>
    <span class="n">option2</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">junction2</span><span class="o">.</span><span class="n">stop</span> <span class="o">-</span> <span class="n">junction1</span><span class="o">.</span><span class="n">start</span><span class="p">)</span> \
        <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_de_novo_exon_length</span>

    <span class="k">if</span> <span class="n">option1</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">junction1</span><span class="o">.</span><span class="n">stop</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">junction2</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span>
    <span class="k">elif</span> <span class="n">option2</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">junction2</span><span class="o">.</span><span class="n">stop</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">junction1</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="bp">False</span><span class="p">,</span> <span class="bp">False</span>

</div>
<div class="viewcode-block" id="ExonJunctionAdjacencies"><a class="viewcode-back" href="../../../outrigger.index.adjacencies.html#outrigger.index.adjacencies.ExonJunctionAdjacencies">[docs]</a><span class="k">class</span> <span class="nc">ExonJunctionAdjacencies</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Annotate junctions with neighboring exons (upstream or downstream)&quot;&quot;&quot;</span>

    <span class="n">exon_types</span> <span class="o">=</span> <span class="s">&#39;exon&#39;</span><span class="p">,</span> <span class="n">NOVEL_EXON</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">metadata</span><span class="p">,</span> <span class="n">db</span><span class="p">,</span> <span class="n">junction_id</span><span class="o">=</span><span class="n">JUNCTION_ID</span><span class="p">,</span>
                 <span class="n">exon_start</span><span class="o">=</span><span class="n">EXON_START</span><span class="p">,</span> <span class="n">exon_stop</span><span class="o">=</span><span class="n">EXON_STOP</span><span class="p">,</span>
                 <span class="n">chrom</span><span class="o">=</span><span class="n">CHROM</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">STRAND</span><span class="p">,</span>
                 <span class="n">max_de_novo_exon_length</span><span class="o">=</span><span class="n">MAX_DE_NOVO_EXON_LENGTH</span><span class="p">,</span>
                 <span class="n">n_jobs</span><span class="o">=-</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Initialize class to get upstream/downstream exons of junctions</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        metadata : pandas.DataFrame</span>
<span class="sd">            A table of splice junctions with the columns indicated by the</span>
<span class="sd">            variables `junction_id`, `exon_start`, `exon_stop`, `chrom`,</span>
<span class="sd">            `strand`</span>
<span class="sd">        db : gffutils.FeatureDB</span>
<span class="sd">            Gffutils Database of gene, transcript, and exon features.</span>
<span class="sd">        junction_id, exon_start, exon_stop, chrom, strand : str</span>
<span class="sd">            Columns in `metadata`</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="n">junction_id</span><span class="p">,</span> <span class="n">exon_start</span><span class="p">,</span> <span class="n">exon_stop</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">strand</span>

        <span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">column</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">metadata</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&#39;The required column {} is not in the splice &#39;</span>
                                 <span class="s">&#39;junction dataframe&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">column</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">junction_id</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">junction_id</span> <span class="o">=</span> <span class="n">junction_id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">exon_start</span> <span class="o">=</span> <span class="n">exon_start</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">exon_stop</span> <span class="o">=</span> <span class="n">exon_stop</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="o">=</span> <span class="n">strand</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">db</span> <span class="o">=</span> <span class="n">db</span>
        <span class="n">progress</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">Looking up which exons are already defined ...&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">existing_exons</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span>
            <span class="n">i</span><span class="p">[</span><span class="s">&#39;id&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span>
                <span class="s">&#39;select id from features where featuretype = &quot;exon&quot;&#39;</span><span class="p">))</span>
        <span class="n">done</span><span class="p">(</span><span class="n">n_tabs</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_de_novo_exon_length</span> <span class="o">=</span> <span class="n">max_de_novo_exon_length</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">n_jobs</span> <span class="o">=</span> <span class="n">n_jobs</span>

<div class="viewcode-block" id="ExonJunctionAdjacencies.detect_exons_from_junctions"><a class="viewcode-back" href="../../../outrigger.index.adjacencies.html#outrigger.index.adjacencies.ExonJunctionAdjacencies.detect_exons_from_junctions">[docs]</a>    <span class="k">def</span> <span class="nf">detect_exons_from_junctions</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Find exons based on gaps in junctions&quot;&quot;&quot;</span>
        <span class="n">junctions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">Region</span><span class="p">),</span>
                              <span class="n">name</span><span class="o">=</span><span class="s">&#39;region&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
        <span class="n">junctions</span> <span class="o">=</span> <span class="n">junctions</span><span class="o">.</span><span class="n">to_frame</span><span class="p">()</span>
        <span class="n">junctions</span><span class="p">[</span><span class="s">&#39;chrom&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">junctions</span><span class="p">[</span><span class="s">&#39;region&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">chrom</span><span class="p">)</span>
        <span class="n">junctions</span><span class="p">[</span><span class="s">&#39;start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">junctions</span><span class="p">[</span><span class="s">&#39;region&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
        <span class="n">junctions</span><span class="p">[</span><span class="s">&#39;stop&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">junctions</span><span class="p">[</span><span class="s">&#39;region&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">stop</span><span class="p">)</span>
        <span class="n">junctions</span><span class="p">[</span><span class="s">&#39;strand&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">junctions</span><span class="p">[</span><span class="s">&#39;region&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">junctions</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s">&#39;chrom&#39;</span><span class="p">):</span>
            <span class="c"># Only get left-adjacent novel exons since there has to be a</span>
            <span class="c"># junction on both sides, and since we iterate over ALL junctions,</span>
            <span class="c"># if we get all left and right exons for all junctions, we&#39;re</span>
            <span class="c"># double-counting exons</span>
            <span class="n">progress</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">Finding all exons on chromosome {chrom} &#39;</span>
                     <span class="s">&#39;...&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">))</span>
            <span class="n">exon_locations</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">joblib</span><span class="o">.</span><span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">n_jobs</span><span class="p">)(</span>
                <span class="n">joblib</span><span class="o">.</span><span class="n">delayed</span><span class="p">(</span><span class="n">_neighboring_exons</span><span class="p">)(</span><span class="n">junction</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="s">&#39;left&#39;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">junction</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">region</span><span class="p">))</span>
            <span class="n">done</span><span class="p">(</span><span class="n">n_tabs</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

            <span class="n">progress</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t\t</span><span class="s">Filtering for only novel exons on chromosome &#39;</span>
                     <span class="s">&#39;{chrom} ...&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">))</span>
            <span class="n">novel_exons</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">exon_locations</span> <span class="k">if</span>
                              <span class="s">&#39;exon:{}:{}-{}:{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="o">*</span><span class="n">x</span><span class="p">)</span>
                              <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">existing_exons</span><span class="p">)</span>
            <span class="n">done</span><span class="p">(</span><span class="n">n_tabs</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

            <span class="n">progress</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t\t</span><span class="s">Creating gffutils.Feature objects for each novel &#39;</span>
                     <span class="s">&#39;exon, plus potentially its overlapping gene&#39;</span><span class="p">)</span>
            <span class="n">exon_features</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">exon_location_to_feature</span><span class="p">(</span><span class="o">*</span><span class="n">x</span><span class="p">)</span>
                             <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">novel_exons</span><span class="p">]</span>
            <span class="n">done</span><span class="p">(</span><span class="n">n_tabs</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

            <span class="n">progress</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t\t</span><span class="s">Updating gffutils database with {n} novel exons on &#39;</span>
                     <span class="s">&#39;chromosome {chrom} ...&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span>
                                                     <span class="n">n</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">novel_exons</span><span class="p">)))</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">exon_features</span><span class="p">,</span>
                               <span class="n">make_backup</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                               <span class="n">id_spec</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;novel_exon&#39;</span><span class="p">:</span> <span class="s">&#39;location_id&#39;</span><span class="p">},</span>
                               <span class="n">transform</span><span class="o">=</span><span class="n">transform</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="n">progress</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">No novel exons found on chromosome &#39;</span>
                         <span class="s">&#39;{chrom}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">))</span>
            <span class="n">done</span><span class="p">(</span><span class="n">n_tabs</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

        <span class="c"># For up to 1000x faster queries, re-Analyze the database now that it</span>
        <span class="c"># has been updated</span>
        <span class="n">maybe_analyze</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="ExonJunctionAdjacencies.exon_location_to_feature"><a class="viewcode-back" href="../../../outrigger.index.adjacencies.html#outrigger.index.adjacencies.ExonJunctionAdjacencies.exon_location_to_feature">[docs]</a>    <span class="k">def</span> <span class="nf">exon_location_to_feature</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">strand</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">STRANDS</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="s">&#39;.&#39;</span>
        <span class="n">overlapping_genes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">region</span><span class="p">(</span><span class="n">seqid</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
                                           <span class="n">end</span><span class="o">=</span><span class="n">stop</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span>
                                           <span class="n">featuretype</span><span class="o">=</span><span class="s">&#39;gene&#39;</span><span class="p">)</span>

        <span class="n">exon_id</span> <span class="o">=</span> <span class="s">&#39;exon:{chrom}:{start}-{stop}:{strand}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span><span class="n">stop</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">)</span>

        <span class="n">attributes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">overlapping_genes</span><span class="p">:</span>
            <span class="n">attributes</span> <span class="o">=</span> <span class="n">merge_attributes</span><span class="p">(</span><span class="n">attributes</span><span class="p">,</span> <span class="n">g</span><span class="o">.</span><span class="n">attributes</span><span class="p">)</span>

        <span class="n">exon</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">Feature</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="n">OUTRIGGER_DE_NOVO</span><span class="p">,</span>
                                <span class="n">featuretype</span><span class="o">=</span><span class="n">NOVEL_EXON</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
                                <span class="n">end</span><span class="o">=</span><span class="n">stop</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">exon_id</span><span class="p">,</span>
                                <span class="n">attributes</span><span class="o">=</span><span class="n">attributes</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">exon</span>
</div>
<div class="viewcode-block" id="ExonJunctionAdjacencies.write_de_novo_exons"><a class="viewcode-back" href="../../../outrigger.index.adjacencies.html#outrigger.index.adjacencies.ExonJunctionAdjacencies.write_de_novo_exons">[docs]</a>    <span class="k">def</span> <span class="nf">write_de_novo_exons</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s">&#39;novel_exons.gtf&#39;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Write all de novo exons to a gtf&quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">novel_exons</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="n">NOVEL_EXON</span><span class="p">,</span>
                                                   <span class="n">order_by</span><span class="o">=</span><span class="n">ORDER_BY</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">noveL_exon</span> <span class="ow">in</span> <span class="n">novel_exons</span><span class="p">:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">noveL_exon</span><span class="p">)</span> <span class="o">+</span> <span class="s">&#39;</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="ExonJunctionAdjacencies.add_exon_to_db"><a class="viewcode-back" href="../../../outrigger.index.adjacencies.html#outrigger.index.adjacencies.ExonJunctionAdjacencies.add_exon_to_db">[docs]</a>    <span class="k">def</span> <span class="nf">add_exon_to_db</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">strand</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">STRANDS</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">overlapping_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">region</span><span class="p">(</span><span class="n">seqid</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
                                                <span class="n">end</span><span class="o">=</span><span class="n">stop</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span>
                                                <span class="n">featuretype</span><span class="o">=</span><span class="s">&#39;gene&#39;</span><span class="p">))</span>

        <span class="n">exon_id</span> <span class="o">=</span> <span class="s">&#39;exon:{chrom}:{start}-{stop}:{strand}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span><span class="n">stop</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">overlapping_genes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">exon</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">Feature</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="n">OUTRIGGER_DE_NOVO</span><span class="p">,</span>
                                    <span class="n">featuretype</span><span class="o">=</span><span class="n">NOVEL_EXON</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
                                    <span class="n">end</span><span class="o">=</span><span class="n">stop</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">exon_id</span><span class="p">)</span>
            <span class="n">progress</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">Added a novel exon ({}), located in an unannotated&#39;</span>
                     <span class="s">&#39; gene&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">exon</span><span class="o">.</span><span class="n">id</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">update</span><span class="p">([</span><span class="n">exon</span><span class="p">],</span> <span class="n">id_spec</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;novel_exon&#39;</span><span class="p">:</span> <span class="s">&#39;location_id&#39;</span><span class="p">},</span>
                           <span class="n">transform</span><span class="o">=</span><span class="n">transform</span><span class="p">)</span>
            <span class="k">return</span>

        <span class="n">de_novo_exons</span> <span class="o">=</span> <span class="p">[</span><span class="n">gffutils</span><span class="o">.</span><span class="n">Feature</span><span class="p">(</span>
            <span class="n">chrom</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="n">OUTRIGGER_DE_NOVO</span><span class="p">,</span> <span class="n">featuretype</span><span class="o">=</span><span class="n">NOVEL_EXON</span><span class="p">,</span>
            <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="n">stop</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">exon_id</span> <span class="o">+</span> <span class="n">g</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
            <span class="n">attributes</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">attributes</span><span class="o">.</span><span class="n">items</span><span class="p">()))</span>
                         <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">overlapping_genes</span><span class="p">]</span>

        <span class="c"># Add all exons that aren&#39;t already there</span>
        <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">de_novo_exons</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">gene_name</span> <span class="o">=</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">exon</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;gene_name&#39;</span><span class="p">])</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">gene_name</span> <span class="o">=</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">exon</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;gene_id&#39;</span><span class="p">])</span>
                    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                        <span class="n">gene_name</span> <span class="o">=</span> <span class="s">&#39;unknown_gene&#39;</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="c"># Check that the non-novel exon doesn&#39;t exist already</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="p">[</span><span class="n">exon_id</span> <span class="o">+</span> <span class="n">exon</span><span class="o">.</span><span class="n">strand</span><span class="p">]</span>
                <span class="k">except</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">FeatureNotFoundError</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">update</span><span class="p">([</span><span class="n">exon</span><span class="p">],</span>
                                   <span class="n">id_spec</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;novel_exon&#39;</span><span class="p">:</span> <span class="s">&#39;location_id&#39;</span><span class="p">},</span>
                                   <span class="n">transform</span><span class="o">=</span><span class="n">transform</span><span class="p">)</span>
                    <span class="n">progress</span><span class="p">(</span>
                        <span class="s">&#39;</span><span class="se">\t</span><span class="s">Added a novel exon ({}) in the gene {} &#39;</span>
                        <span class="s">&#39;({})&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="n">exon</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">exon</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;gene_id&#39;</span><span class="p">]),</span>
                            <span class="n">gene_name</span><span class="p">))</span>
            <span class="k">except</span> <span class="n">sqlite3</span><span class="o">.</span><span class="n">IntegrityError</span><span class="p">:</span>
                <span class="k">continue</span>
</div>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_single_junction_exon_triple</span><span class="p">(</span><span class="n">direction_ind</span><span class="p">,</span> <span class="n">direction</span><span class="p">,</span> <span class="n">exon_id</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Create exon, direction, junction triples for an exon + its junctions</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        direction_ind : pandas.Series</span>
<span class="sd">            A boolean series of the indices of the junctions matching with the</span>
<span class="sd">            provided exon. The index of the series must be the junction ID</span>
<span class="sd">        direction : str</span>
<span class="sd">            The direction of the exon relative to the junctions, either</span>
<span class="sd">            &quot;upstream&quot; or &quot;downstream&quot;</span>
<span class="sd">        exon_id : str</span>
<span class="sd">            Unique identifier of the exon</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        triples : pandas.DataFrame</span>
<span class="sd">            A (n, 3) sized dataframe of an exon and its adjacent junctions</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">length</span> <span class="o">=</span> <span class="n">direction_ind</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

        <span class="n">exons</span> <span class="o">=</span> <span class="p">[</span><span class="n">exon_id</span><span class="p">]</span> <span class="o">*</span> <span class="n">length</span>
        <span class="n">directions</span> <span class="o">=</span> <span class="p">[</span><span class="n">direction</span><span class="p">]</span> <span class="o">*</span> <span class="n">length</span>
        <span class="n">junctions</span> <span class="o">=</span> <span class="n">direction_ind</span><span class="p">[</span><span class="n">direction_ind</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">exons</span><span class="p">,</span> <span class="n">directions</span><span class="p">,</span> <span class="n">junctions</span><span class="p">)),</span>
                            <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;exon&#39;</span><span class="p">,</span> <span class="s">&#39;direction&#39;</span><span class="p">,</span> <span class="s">&#39;junction&#39;</span><span class="p">])</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_to_stranded_transcript_adjacency</span><span class="p">(</span><span class="n">adjacent_in_genome</span><span class="p">,</span> <span class="n">strand</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;If negative strand, swap the upstream/downstream adjacency</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        adjacent_in_genome : dict</span>
<span class="sd">            dict of two keys, &quot;upstream&quot; and &quot;downstream&quot;, mapping to a boolean</span>
<span class="sd">            series indicating whether the junction is upstream or downstream of</span>
<span class="sd">            a particular exon</span>
<span class="sd">        strand : &quot;-&quot; | &quot;+&quot;</span>
<span class="sd">            Positive or negative strand</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s">&#39;+&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">{</span><span class="n">UPSTREAM</span><span class="p">:</span> <span class="n">adjacent_in_genome</span><span class="p">[</span><span class="n">UPSTREAM</span><span class="p">],</span>
                    <span class="n">DOWNSTREAM</span><span class="p">:</span> <span class="n">adjacent_in_genome</span><span class="p">[</span><span class="n">DOWNSTREAM</span><span class="p">]}</span>
        <span class="k">elif</span> <span class="n">strand</span> <span class="o">==</span> <span class="s">&#39;-&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">{</span><span class="n">UPSTREAM</span><span class="p">:</span> <span class="n">adjacent_in_genome</span><span class="p">[</span><span class="n">DOWNSTREAM</span><span class="p">],</span>
                    <span class="n">DOWNSTREAM</span><span class="p">:</span> <span class="n">adjacent_in_genome</span><span class="p">[</span><span class="n">UPSTREAM</span><span class="p">]}</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c"># If strand is unknown, put both upstream and downstream for each</span>
            <span class="c"># side</span>
            <span class="n">adjacent</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">adjacent_in_genome</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
            <span class="k">return</span> <span class="p">{</span><span class="n">UPSTREAM</span><span class="p">:</span> <span class="n">adjacent</span><span class="p">,</span> <span class="n">DOWNSTREAM</span><span class="p">:</span> <span class="n">adjacent</span><span class="p">}</span>

    <span class="k">def</span> <span class="nf">_junctions_genome_adjacent_to_exon</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">exon</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get indices of junctions next to an exon, in genome coordinates&quot;&quot;&quot;</span>
        <span class="n">chrom_ind</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">chrom</span><span class="p">]</span> <span class="o">==</span> <span class="n">exon</span><span class="o">.</span><span class="n">chrom</span>

        <span class="n">strand_ind</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">strand</span><span class="p">]</span> <span class="o">==</span> <span class="n">exon</span><span class="o">.</span><span class="n">strand</span>
        <span class="n">common_ind</span> <span class="o">=</span> <span class="n">chrom_ind</span> <span class="o">&amp;</span> <span class="n">strand_ind</span>

        <span class="n">upstream_in_genome</span> <span class="o">=</span> \
            <span class="n">common_ind</span> <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">exon_stop</span><span class="p">]</span> <span class="o">==</span> <span class="n">exon</span><span class="o">.</span><span class="n">stop</span><span class="p">)</span>
        <span class="n">downstream_in_genome</span> <span class="o">=</span> \
            <span class="n">common_ind</span> <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">exon_start</span><span class="p">]</span> <span class="o">==</span> <span class="n">exon</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">UPSTREAM</span><span class="p">:</span> <span class="n">upstream_in_genome</span><span class="p">,</span> <span class="n">DOWNSTREAM</span><span class="p">:</span> <span class="n">downstream_in_genome</span><span class="p">}</span>

<div class="viewcode-block" id="ExonJunctionAdjacencies.junctions_adjacent_to_this_exon"><a class="viewcode-back" href="../../../outrigger.index.adjacencies.html#outrigger.index.adjacencies.ExonJunctionAdjacencies.junctions_adjacent_to_this_exon">[docs]</a>    <span class="k">def</span> <span class="nf">junctions_adjacent_to_this_exon</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">exon</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get junctions adjacent to this exon</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        exon : gffutils.Feature</span>
<span class="sd">            An item in a gffutils database</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">dfs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">adjacent_in_genome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_junctions_genome_adjacent_to_exon</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>
        <span class="n">adjacent_in_transcriptome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_to_stranded_transcript_adjacency</span><span class="p">(</span>
            <span class="n">adjacent_in_genome</span><span class="p">,</span> <span class="n">exon</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>

        <span class="n">exon_id</span> <span class="o">=</span> <span class="n">exon</span><span class="o">.</span><span class="n">id</span>
        <span class="k">for</span> <span class="n">direction</span><span class="p">,</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">adjacent_in_transcriptome</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">ind</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
                <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_single_junction_exon_triple</span><span class="p">(</span><span class="n">ind</span><span class="p">,</span> <span class="n">direction</span><span class="p">,</span> <span class="n">exon_id</span><span class="p">)</span>
                <span class="n">dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">dfs</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
</div>
<div class="viewcode-block" id="ExonJunctionAdjacencies.upstream_downstream_exons"><a class="viewcode-back" href="../../../outrigger.index.adjacencies.html#outrigger.index.adjacencies.ExonJunctionAdjacencies.upstream_downstream_exons">[docs]</a>    <span class="k">def</span> <span class="nf">upstream_downstream_exons</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get upstream and downstream exons of each junction</span>

<span class="sd">        The &quot;upstream&quot; and &quot;downstream&quot; is relative to the **junction**, e.g.</span>

<span class="sd">            exonA   upstream      junctionX</span>
<span class="sd">            exonB   downstream    junctionX</span>

<span class="sd">        should be read as &quot;exonA is upstream of juction X&quot; and &quot;exonB is</span>
<span class="sd">        downstream of junctionX&quot;</span>

<span class="sd">        Use junctions defined in ``sj_metadata`` and exons in ``db`` to</span>
<span class="sd">        create triples of (exon, direction, junction), which are read like</span>
<span class="sd">        (subject, object, verb) e.g. (&#39;exon1&#39;, &#39;upstream&#39;, &#39;junction12&#39;), for</span>
<span class="sd">        creation of a graph database.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sj_metadata : pandas.DataFrame</span>
<span class="sd">            A splice junction metadata dataframe with the junction id as the</span>
<span class="sd">            index, with  columns defined by variables ``exon_start`` and</span>
<span class="sd">            ``exon_stop``.</span>
<span class="sd">        db : gffutils.FeatureDB</span>
<span class="sd">            A database of gene annotations created by gffutils. Must have</span>
<span class="sd">            features of type &quot;exon&quot;</span>
<span class="sd">        exon_start : str, optional</span>
<span class="sd">            Name of the column in sj_metadata corresponding to the start of the</span>
<span class="sd">            exon</span>
<span class="sd">        exon_stop : str, optional</span>
<span class="sd">            Name of the column in sj_metadata corresponding to the end of the</span>
<span class="sd">            exon</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        junction_exon_triples : pandas.DataFrame</span>
<span class="sd">            A three-column dataframe describing the relationship of where an</span>
<span class="sd">            exon is relative to junctions</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">n_exons</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">exon_types</span><span class="p">))</span>

        <span class="n">dfs</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">progress</span><span class="p">(</span><span class="s">&#39;Starting annotation of all junctions with known &#39;</span>
                 <span class="s">&#39;neighboring exons ...&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">exon</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">exon_types</span><span class="p">)):</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">10000</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">progress</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">{}/{} exons completed&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n_exons</span><span class="p">))</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">junctions_adjacent_to_this_exon</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>
            <span class="n">dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
        <span class="n">junction_exon_triples</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">dfs</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="n">done</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">junction_exon_triples</span></div></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">BED annotator 0.2.9 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Olga Botvinnik.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>