<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>outrigger.validate.check_splice_sites &#8212; Outrigger 1.0.0 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '1.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="top" title="Outrigger 1.0.0 documentation" href="../../../index.html" />
    <link rel="up" title="Module code" href="../../index.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for outrigger.validate.check_splice_sites</h1><div class="highlight"><pre>
<span></span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">OrderedDict</span>

<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">pybedtools</span>

<span class="n">NT</span> <span class="o">=</span> <span class="mi">2</span>

<span class="n">MAMMALIAN_SPLICE_SITES</span> <span class="o">=</span> <span class="s1">&#39;GT/AG,GC/AG,AT/AC&#39;</span>


<div class="viewcode-block" id="splice_site_str_to_tuple"><a class="viewcode-back" href="../../../outrigger.validate.check_splice_sites.html#outrigger.validate.check_splice_sites.splice_site_str_to_tuple">[docs]</a><span class="k">def</span> <span class="nf">splice_site_str_to_tuple</span><span class="p">(</span><span class="n">splice_site</span><span class="p">):</span>
    <span class="n">pairs</span> <span class="o">=</span> <span class="n">splice_site</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">pairs</span><span class="p">)</span></div>


<div class="viewcode-block" id="maybe_read_chromsizes"><a class="viewcode-back" href="../../../outrigger.validate.check_splice_sites.html#outrigger.validate.check_splice_sites.maybe_read_chromsizes">[docs]</a><span class="k">def</span> <span class="nf">maybe_read_chromsizes</span><span class="p">(</span><span class="n">genome</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">chromsizes</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">chrom</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                <span class="n">size</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">size</span><span class="p">)</span>
                <span class="n">chromsizes</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">size</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">OSError</span><span class="p">:</span>
        <span class="n">chromsizes</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">chromsizes</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">chromsizes</span></div>


<div class="viewcode-block" id="read_splice_sites"><a class="viewcode-back" href="../../../outrigger.validate.check_splice_sites.html#outrigger.validate.check_splice_sites.read_splice_sites">[docs]</a><span class="k">def</span> <span class="nf">read_splice_sites</span><span class="p">(</span><span class="n">bed</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">fasta</span><span class="p">,</span> <span class="n">direction</span><span class="o">=</span><span class="s1">&#39;upstream&#39;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Read splice sites of an exon</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bed : pybedtools.BedTool | str</span>
<span class="sd">        Exons whose splice sites you&#39;re interested in</span>
<span class="sd">    genome : str</span>
<span class="sd">        Location of a chromosome sizes file for the genome</span>
<span class="sd">    fasta : str</span>
<span class="sd">        Location of the genome fasta file</span>
<span class="sd">    direction : &#39;upstream&#39; | &#39;downstream&#39;</span>
<span class="sd">        Which direction of splice sites you want for the exon</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">bed</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">bed</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">bed</span><span class="p">)</span>

    <span class="n">genome</span> <span class="o">=</span> <span class="n">maybe_read_chromsizes</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">direction</span> <span class="o">==</span> <span class="s1">&#39;upstream&#39;</span><span class="p">:</span>
        <span class="n">left</span> <span class="o">=</span> <span class="n">NT</span>
        <span class="n">right</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">elif</span> <span class="n">direction</span> <span class="o">==</span> <span class="s1">&#39;downstream&#39;</span><span class="p">:</span>
        <span class="n">left</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">right</span> <span class="o">=</span> <span class="n">NT</span>

    <span class="n">flanked</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">flank</span><span class="p">(</span><span class="n">l</span><span class="o">=</span><span class="n">left</span><span class="p">,</span> <span class="n">r</span><span class="o">=</span><span class="n">right</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="n">genome</span><span class="p">)</span>
    <span class="n">seqs</span> <span class="o">=</span> <span class="n">flanked</span><span class="o">.</span><span class="n">sequence</span><span class="p">(</span><span class="n">fi</span><span class="o">=</span><span class="n">fasta</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">seqs</span><span class="o">.</span><span class="n">seqfn</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">records</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
        <span class="n">records</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">records</span><span class="p">],</span>
                            <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="n">b</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="n">bed</span><span class="p">])</span>
    <span class="c1"># import pdb; pdb.set_trace()</span>
    <span class="k">return</span> <span class="n">records</span></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">Outrigger</a></h1>


<a href="../../../index.html">
<img src="../../../_static/logo-1x.png" width=80%>
</a>

<h3><a href="../../../index.html">Table Of Contents</a></h3>

<ul>
  <li><a href="../../../index.html">Home</a></li>
  <li><a href="../../../contents.html">Contents</a></li>
  <li><a href="../../../installation.html">Install</a></li>
  <li><a href="../../../Usage.html">Usage</a></li>
  <li><a href="../../../subcommands/outrigger_index.html"><code>index</code>: Detect exons</a></li>
  <li><a href="../../../subcommands/outrigger_validate.html"><code>validate</code>: Remove non-canonical splice sites</a></li>
  <li><a href="../../../subcommands/outrigger_psi.html"><code>psi</code>: Quantify exon inclusion</a></li>
  <li><a href="../../../history.html">Changelog</a></li>
  <li><a href="../../../license.html">License</a></li>
</ul>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015-2017, Olga Botvinnik.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.4.8</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.9</a>
      
    </div>

    

    
  </body>
</html>